File: backend/src/models/book.ts
import mongoose from "mongoose";
import { title } from "process";

const BookSchema = new mongoose.Schema({
    title: { type: String, required: true, trim: true },
    author: { type: String, required: true, trim: true},
    category: { type: String, trim: true },
    genre: { type: String, trim: true },
    publicationDate: { type: String, required: true },
    isbn: {
        type: String,
        required: true,
        validate: {
            validator: function (val: any) {
                // Ensures that at least one of the ISBN fields has a value
                return /^\d{9}$|^\d{10}$|^\d{13}$/.test(val);
            },
            message: 'At least one of SBN, ISBN10, or ISBN13, must be provided in the ISBN object.',
        },
    },
    quantity: { 
        type: String,
        required: true,
        validate: {
            validator: function (val: any) {
                // Ensures that at least one of the ISBN fields has a value
                return /^[1-9][0-9]{0,1000000000}$/.test(val);
            },
            message: 'The number of copies must be a number.',
        },
    },
    description: { type: String, trim: true },
    language: { type: String, trim: true },
    format: { type: String, trim: true },
    publisher: { type: String, trim: true },
    bookCover: { type: String, trim: true },
});

BookSchema.index({ title: 1 });
BookSchema.index({ author: 1 });
BookSchema.index({ genre: 1 });
BookSchema.index({ isbn: 1 });

export const Book = mongoose.model('Book', BookSchema);

File: frontend/lib/models/book.dart
class Book {
  final String author;
  final String title;
  final String publicationDate;
  final String quantity;
  final String isbn;
  final String? category;
  final String? genre;
  final String? description;
  final String? language;
  final String? format;
  final String? publisher;
  final String? bookCover;

  Book({
    required this.author,
    required this.title,
    required this.publicationDate,
    required this.quantity,
    required this.isbn,
    this.category,
    this.genre,
    this.description,
    this.language,
    this.format,
    this.publisher,
    this.bookCover,
  });

  Map<String, dynamic> toJson() {
    return {
      'author': author,
      'title': title,
      'publicationDate': publicationDate,
      'quantity': quantity,
      'isbn': isbn,
      'category': category,
      'genre': genre,
      'description': description,
      'language': language,
      'format': format,
      'publisher': publisher,
      'bookCover': bookCover,
    };
  }

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      author: json['author'],
      title: json['title'],
      publicationDate: json['publicationDate'],
      quantity: json['quantity'],
      isbn: json['isbn'],
      category: json['category'],
      genre: json['genre'],
      description: json['description'],
      language: json['language'],
      format: json['format'],
      publisher: json['publisher'],
      bookCover: json['bookCover'],
    );
  }
}